### PR TITLE
Add linalg.vander

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #76303

This PR adds `linalg.vander`, the linalg version of `torch.vander`.

We add autograd support and support for batched inputs.

We also take this chance to improve the docs (TODO: Check that they
render correctly!) and add an OpInfo.

**Discussion**: The current default for the `increasing` kwargs is extremely
odd as it is the opposite of the classical definition (see [wiki](https://en.wikipedia.org/wiki/Vandermonde_matrix)). This is
reflected in the docs, where I explicit both the odd defaults that we
use and the classical definition. See also [this stackoverflow post](https://stackoverflow.com/a/71758047/5280578), which 
shows how people are confused by this defaults.

My take on this would be to correct the default to be `increasing=True`
and document the divergence with NumPy (as we do for other `linalg`
functions) as:

- It is what people expect
- It gives the correct determinant called "the [Vandermonde determinant](https://en.wikipedia.org/wiki/Vandermonde_polynomial)" rather than (-1)^{n-1} times that quantity (ugh).
- [Minor] It is more efficient (no `flip` needed)
- Since it's under `linalg.vander`, it's strictly not a drop-in replacement for `np.vander`.

We will deprecate `torch.vander` in a PR after this one in this stack
(once we settle on what's the correct default).

Thoughts? @mruberry

cc @kgryte @rgommers 

Fixes https://github.com/pytorch/pytorch/issues/60197

### CODE REVIEW

**Comment by rgommers** [comment]
I did a bit of archeology, this behavior in NumPy was discussed around 2008 and seems to have been inherited from Matlab:
- https://mail.python.org/archives/list/numpy-discussion@python.org/thread/NJDGY7GYGKWSDLSX4ODZ5HGZ7KM3YNDC/#NJDGY7GYGKWSDLSX4ODZ5HGZ7KM3YNDC
- https://mail.python.org/archives/list/numpy-discussion@python.org/thread/NL6X7CYOQNQIF75VZUDGY5DVTGKOPABA/#NL6X7CYOQNQIF75VZUDGY5DVTGKOPABA

The `increasing=False` keyword was added in 2014 to allow people to switch to the more conventional definition in the literature. For backwards compat it could not be changed at that point. It looks like the option of deprecating `np.vander` and introducing `np.linalg.vander` was not discussed. 

The choice made in this PR of adding `torch.linalg.vander` with the behavior that is conventional in the literature makes sense to me. I do think there's then no reason for adding an `increasing` keyword at all. How about leaving it out and just documenting instead the one-liner you need to match `np.vander`?

Also, we may want to raise this again with NumPy, the function could be moved to `numpy.linalg` with changed behavior.

File: torch/linalg/__init__.py [hunk 1]
```diff
@@ -2378,3 +2378,63 @@
     >>> torch.dist(Q.mT @ Q, torch.eye(4))
     tensor(6.2158e-07)
 """)
+
+vander = _add_docstr(_linalg.linalg_vander, r"""
+vander(x, N=None, increasing=False) -> Tensor
```
Comment (rgommers) [inline torch/linalg/__init__.py:2383]: This should have `increasing=True`

**Comment by kgryte** [comment]
I'd also recommend leaving out the `increasing` kwarg.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -4039,4 +4039,39 @@ Tensor linalg_solve_triangular(
...
+  auto dtype = x_.scalar_type();
+  auto options = x_.options()
+                   .dtype(at::isIntegralType(dtype, /*includeBool=*/true) ? kLong : dtype);
+
+  // n = 0 or n = 1 case (Empty or one row case)
+  shape.push_back(std::min<int64_t>(n, 1LL));
+  auto ones =  x_.new_ones(shape, options);
+
+  if (n <= 1) {
+    // new_ones does not propagate requires_grad. UGH
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:4062]: Even thought the tensor options should have `requires_grad=True`?

File: torch/linalg/__init__.py (hunk 1 above)
Comment (mruberry) [reply torch/linalg/__init__.py:2383]: (or remove the increasing argument all together)

File: torch/linalg/__init__.py [hunk 3]
```diff
@@ -2378,3 +2378,63 @@
...
+
+If :attr:`increasing`\ `= False`, the powers of the columns are returned in decreasing order.
+
+If :attr:`N`\ `= None`, then `N = x.size(-1)` so that the output is a square matrix.
+
+Supports inputs of all dtypes.
+Also supports batches of vectors, and if :attr:`x` is a batch of vectors then
+the output has the same batch dimensions.
+
+Differences with `numpy.vander`:
```
Comment (mruberry) [inline torch/linalg/__init__.py:2407]: Maybe put this in a note?

File: torch/linalg/__init__.py [hunk 4]
```diff
@@ -2378,3 +2378,63 @@
...
+Keyword args:
+    N (int, optional): Number of columns in the output. Default: `x.size(-1)`
+    increasing (bool, optional): Order of the powers of the columns.
+        If `True`, the powers increase from left to right.
+        The order is reversed otherwise.
+        Default: `True`
+
+Example::
+
+    >>> x = torch.tensor([1, 2, 3, 5])
```
Comment (mruberry) [inline torch/linalg/__init__.py:2424]: Nice examples

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -3136,6 +3136,37 @@ def sample_inputs_linalg_cond(op_info, device, dtype, requires_grad=False, **kwa
     for shape in shapes:
         yield SampleInput(make_arg(shape))
 
+def sample_inputs_linalg_vander(op_info, device, dtype, requires_grad=False, **kwargs):
+    make_arg = partial(make_tensor, dtype=dtype, device=device, requires_grad=requires_grad)
+
+    # autograd is not supported for inputs with zero number of elements
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:3142]: Those inputs can be added to the samples on when `requires_grad=False`

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -3136,6 +3136,37 @@ def sample_inputs_linalg_cond(op_info, device, dtype, requires_grad=False, **kwa
...
+        yield SampleInput(make_arg(shape), kwargs=dict(increasing=increasing))
+        n = shape[-1] if len(shape) > 0 else 1
+        for i in range(3):
+            # n-1, n, n+1
+            N = n - (i - 1)
+            if N < 0:
+                continue
+            yield SampleInput(make_arg(shape), kwargs=dict(N=N, increasing=increasing))
+
+def np_vander_batched(x, N=None, increasing=True):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:3157]: Nice reference operation

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -10650,6 +10681,15 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
            dtypes=floating_and_complex_types(),
            sample_inputs_func=sample_inputs_linalg_slogdet,
            decorators=[skipCUDAIfNoMagma, skipCPUIfNoLapack],),
+    OpInfo('linalg.vander',
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:10684]: This looks good but if we don't want to add an additional test then we probably want to add "reference inputs" for vander that test it vs. the NumPy reference on a wider range of inputs. The reference inputs typically yield from the sample inputs and then have a list of additional cases. This will autogenerate a test where those inputs are passed to PyTorch's implementation and the NumPy reference implementation and the results compared.

**Comment by mruberry** [comment]
I'm not sure what's going on with the public bindings -- maybe @albanD knows?

Overall this looks great. re: `increasing` I can see including it just as an option for NumPy users to get the historic NumPy behavior, but if it's a pain to support I can also see removing it entirely.

File: torch/linalg/__init__.py (hunk 4 above)
Comment (Lezcano) [reply torch/linalg/__init__.py:2424]: Shamelessly nicked from the docs from `torch.vander`.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 2 above)
Comment (Lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:4062]: `tensor.options()` does not propagate `requires_grad` (??!!).

File: torch/linalg/__init__.py (hunk 3 above)
Comment (mruberry) [reply torch/linalg/__init__.py:2407]: nbd

File: torch/testing/_internal/common_methods_invocations.py (hunk 7 above)
Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:10684]: I don't know which other cases we could want to test this function on, given that we already have a test for non-contiguous samples and so on. I think that the set of samples (12) are fairly comprehensive.

File: torch/testing/_internal/common_methods_invocations.py (hunk 5 above)
Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:3142]: Actually that's a stale comment. I added an input with shape `()`.

**Comment by Lezcano** [comment]
Compiling the ideas from the discussion above I resolved:
- To remove the `increasing` flag and leave `increasing=True` as the default behaviour.
- To document an alternative to it (`linalg.vander(x, N).flip(-1)`) in the "differences with NumPy" section of the docs.

File: torch/linalg/__init__.py [hunk 8]
```diff
@@ -2378,3 +2378,53 @@
...
+    tensor([[  1,   1,   1,   1],
+            [  1,   2,   4,   8],
+            [  1,   3,   9,  27],
+            [  1,   5,  25, 125]])
+    >>> linalg.vander(x, N=3)
+    tensor([[ 1,  1,  1],
+            [ 1,  2,  4],
+            [ 1,  3,  9],
+            [ 1,  5, 25]])
+""")
```
Comment (albanD) [inline torch/linalg/__init__.py:2430]: You can add the same as https://github.com/pytorch/pytorch/blob/f02b7a9c36dd6182da694bc47a5c345285dfd951/torch/nn/functional.py#L4328-L4329 here to avoid the failure.
It would be great when we have a minute though to fix the C binding directly!

Comment (Lezcano) [reply torch/linalg/__init__.py:2430]: As per the the discussion in the ldl PR, I'm adding it to the list of exceptions.

**Comment by Lezcano** [comment]
I fixed everything. If the CI is green and @mruberry agrees, this should be ready to go.

**Review by albanD — APPROVED** [review]
SGTM

**Comment by Lezcano** [comment]
Uf, sorry, I had never seen that comprehensive test. Will test again tomorrow.

**Comment by malfet** [comment]
@Lezcano looks like a land race to me, but also please double check rocm builds before re-landing

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 2 above)
Comment (ngimel) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:4061]: What are we trying to achieve with this `requires_gradness`? Gradient won't be propagated back to x anyway, because there's no connection between new_ones and x.

Comment (ngimel) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:4061]: One (not particularly great) way of making sure that x gets 0 derivative in this case is to construct output as `x * 0 + 1` with some slicing to make sure sizes are correct.

Comment (nikitaved) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:4061]: FWIW, `ones_like` does propagate grads.

Comment (nikitaved) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:4061]: I agree with @ngimel, the all-ones column is always ones no matter what the input is, so there is no need need in making it require grad.

Comment (ngimel) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:4061]: We might want the gradient chain not break and make `x` receive all zero gradients, but I don't think what we have here accomplishes that.

**Comment by ngimel** [comment]
Looks like comprehensive test tests all the dtypes listed in the OpInfo on all the sample inputs (unlike test_dtypes), and cumprod is indeed not implemented for cpu low precision type (nor should it be).

**Comment by mruberry** [comment]
> Looks like comprehensive test tests all the dtypes listed in the OpInfo on all the sample inputs (unlike test_dtypes), and cumprod is indeed not implemented for cpu low precision type (nor should it be).

Note that test_dtypes does test all dtypes and all sample inputs but it just reports if some but not all samples are supported because sometimes ops only have partial support for a dtype (which is lame but oh well).

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 2 above)
Comment (nikitaved) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:4061]: I see your point. Then we could select a single element from the tensor with, say, `index`, multiply it with 0, add 1 and then expand to the shape of `x`. Should be better than multiplying the whole `x` with zero.

Comment (Lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:4061]: I needed this because otherwise gradchecks would complain that "no output is differentiable". I'm happy to hear waht's a good fix for this. thoughts @albanD?

**Comment by Lezcano** [comment]
Which types do you think vander should support? @ngimel @mruberry I don't have any strong opinions, and I just went with the more general set I thought I could find

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 2 above)
Comment (nikitaved) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:4061]: What about this?

```python
In [1]: import torch

In [2]: x = torch.rand(3, 3, requires_grad=True)

In [3]: y = x.pow(0.0)

In [4]: y.sum().backward()

In [5]: x.grad
Out[5]: 
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])

```

Comment (ngimel) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:4061]: my favorite is `y=x*0+1`, but both @nikitaved's and my option will require some slicing to make sure output is correct size. If we didn't support batch dimension it would be easier :-) Also, both options suffer from unexpected nan propagation, so maybe do `y=(x*0).fill_(1)` which is a bit of perf penalty but oh well.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 9]
```diff
@@ -4039,4 +4039,33 @@ Tensor linalg_solve_triangular(
...
+    ones.requires_grad_(x_.requires_grad());
+    return ones;
+  } else {
+    // Append cumprod of the oher 0...n-1 powers
```
Comment (nikitaved) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:4065]: nit: typo and the powers I guess are `1,...,n-1`?

**Comment by mruberry** [comment]
> Which types do you think vander should support? @ngimel @mruberry I don't have any strong opinions, and I just went with the more general set I thought I could find

The dtypes that torch.vander supports seem like a reasonable set to start with?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 2 above)
Comment (nikitaved) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:4061]: I think this idea could be potentially improved perf-wise if done on just a single element of `x` with proper `expand`.

Comment (albanD) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:4061]: Well, on paper a "detached" graph is a valid way to represent a Tensor full of zeros.
It's just that we don't usually gradcheck functions that are independent of their input.

If we really want to do this, you can re-use some test function we have for gradcheck: `output = torch._C._functions.UndefinedGrad()(input)`. This function will just ignore the input and return undefined gradients (so gradients of zeros).
Note that not keeping the graph around is even better for perf though!

Comment (Lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:4061]: Ok, let's just not check those inputs.

Comment (Lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:4061]: Actually just setting `requires_grad=False` for those inputs does not work. I'm tentatively leaving it as it is, as:
- It passes the tests
- If you are using this function to generate a tensor full of ones, you have larger issues to worry about than the fact that we are tracking the derivatives for that leaf tensor you'll never use.

Comment (ngimel) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:4646]: But this is wrong, this is creating a dangling tensor with requires_grad that breaks the gradient propagation. Can we skip the tests instead? vander with n=1 is essentially non-differentiable and should be treated as such.

Comment (ngimel) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:4646]: Getting 0 gradient is different from not getting gradient (the chain is not broken in the former, and broken in the latter). In vander with n=1 we have the latter (x.grad will be None, if x is not leaf all nodes below x won't get gradients). I agree that vander should not be used to get tensor full of ones, but having workarounds in the code just to pacify the tests (and not have the "right" behavior, for however right is defined) is not the way. Returning ones that don't require grad is fine with me, and it should be fine with the tests.  Or it should be commented in the code "we know this is wrong and silly, but that's the easiest thing to do and there's no harm".

Comment (Lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:4646]: After syncing with @albanD, we decided the simplest solution is to just disallow `N <= 1` for now. If a user asks for it, we'll add it then taking into account what they think it's a good design.

**Comment by Lezcano** [comment]
@mruberry Just pushed a simpler version where `N > 1`. I also narrowed down the dtypes that the function accepts. I updated the docs accordingly. Could you have another look?

File: torch/linalg/__init__.py [hunk 10]
```diff
@@ -2527,3 +2527,54 @@
...
+If :attr:`N`\ `= None`, then `N = x.size(-1)` so that the output is a square matrix.
+
+Supports inputs of float, double, cfloat, cdouble, and integral dtypes.
+Also supports batches of vectors, and if :attr:`x` is a batch of vectors then
+the output has the same batch dimensions.
+
+Differences with `numpy.vander`:
+
+- Unlike `numpy.vander`, this function returns the powers of :attr:`x` in ascending order.
+  To get them in the reverse order call ``linalg.vander(x, N).flip(-1)``.
```
Comment (mruberry) [inline torch/linalg/__init__.py:2558]: Nice use of flip!

File: torch/testing/_internal/common_methods_invocations.py (hunk 7 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:11340]: OK

**Review by mruberry — APPROVED** [review]
Way to crush it, @Lezcano!

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 2 above)
Comment (ezyang) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: This happened again with composite compliance for linalg.matrix_power, I would quite like it if we had a consolidated strategy for this
